syntax = "proto3";

package mesh;

import "google/protobuf/struct.proto";
import "google/protobuf/timestamp.proto";

option go_package = "github.com/groovy-byte/agent-mesh-core/proto";

// Defines the core data structures and service contracts for the Adaptive OS Mesh.

// --- OS Resource Tracking ---

message OSResources {
  double cpu_usage_percent = 1;
  uint64 memory_used_bytes = 2;
  uint64 memory_total_bytes = 3;
  double disk_io_wait = 4;
}

// --- Mesh Control Plane ---

enum AgentRole {
  OPERATIONAL = 0; // Low-latency, high-throughput tasks via NATS.
  STRATEGIC = 1;   // Complex, reasoning-heavy tasks via gRPC.
}

message HandshakeRequest {
  string agent_id = 1;
  repeated string capabilities = 2;
  AgentRole initial_role = 3;
}

message HandshakeResponse {
  string session_id = 1;
  bool approved = 2;
  string error_message = 3;
  OSResources resource_limits = 4;
}

message Heartbeat {
  string agent_id = 1;
  google.protobuf.Timestamp timestamp = 2;
  OSResources current_load = 3;
  AgentRole current_role = 4;
}

// --- Agent Actions & Responses ---

message AgentAction {
  string agent_id = 1;
  string action_type = 2; // e.g., "OS_COMPILATION", "REASONING", "SEARCH".
  OSResources resource_impact = 3;
  google.protobuf.Struct payload = 4; // Flexible payload for arbitrary task data.
  string reasoning_chain = 5; // Context for strategic tasks.
  string task_intent = 6; // Intent declaration for predictive resource management.
  uint64 data_size_bytes = 7; // Size hint for hardware-aware scheduling.
}

message ActionResponse {
  bool success = 1;
  google.protobuf.Struct result = 2;
  string error = 3;
  bool promotion_suggested = 4; // Suggests a role change to STRATEGIC.
  string routing_provider = 5; // The hardware path chosen by the scheduler (e.g., "CPU_AVX2", "GPU_CUDA").
  AgentRole required_role = 6; // Enforced role from the controller.
}

// --- Hardware-Aware Inference ---

message InferenceRequest {
  string agent_id = 1;
  string prompt = 2;
  uint32 max_tokens = 3;
  float temperature = 4;
  uint64 expected_kv_cache_bytes = 5; // Hint for memory-intensive inference tasks.
}

message InferenceResponse {
  string text = 1;
  uint32 tokens_used = 2;
  string hardware_path = 3; 
  float latency_ms = 4;
  float throughput_gbs = 5; 
  bool avx512_usage = 6;
}

// --- Synthesis Protocol ---

message SynthesisRequest {
  repeated string agent_ids = 1;
  string target_goal = 2;
  repeated AgentAction actions_to_merge = 3;
}

message SynthesisResponse {
  string synthesized_state = 1;
  float confidence_score = 2;
}

// --- Strategic Service (gRPC) ---

service StrategicMesh {
  // Registers an agent with the mesh.
  rpc RegisterAgent(HandshakeRequest) returns (HandshakeResponse);
  
  // Executes a high-complexity task requiring strategic planning.
  rpc ExecuteStrategicAction(AgentAction) returns (ActionResponse);
  
  // Performs a semantic search over the knowledge base.
  rpc SemanticSearch(SearchRequest) returns (SearchResponse);

  // Retrieves the last known state for a failed agent to allow for recovery.
  rpc GetStateReconstitution(HandshakeRequest) returns (AgentAction);

  // Merges and synthesizes outputs from multiple agents.
  rpc SynthesizeOutputs(SynthesisRequest) returns (SynthesisResponse);

  // Executes a hardware-aware inference request.
  rpc GenerateResponse(InferenceRequest) returns (InferenceResponse);

  // Retrieves performance and audit statistics for the mesh.
  rpc GetMeshStats(StatsRequest) returns (MeshStats);
}

// --- Auditing & Statistics ---

message StatsRequest {}

message MeshStats {
  int32 agents_active = 1;
  map<string, AgentMetrics> agent_logs = 2;
  map<string, InfluenceMap> contribution_matrix = 3;
}

message AgentMetrics {
  uint32 tool_calls = 1;
  repeated string failed_tasks = 2;
  float avg_latency_ms = 3;
  uint32 total_tokens = 4;
}

message InfluenceMap {
  map<string, double> influence = 1;
}

// --- Search Protocol ---

message SearchRequest {
  string agent_id = 1;
  string query = 2;
  int32 max_results = 3;
}

message SearchResponse {
  repeated SearchResult results = 1;
  string reasoning_context = 2; 
}

message SearchResult {
  string source = 3; // The origin of the search result (e.g., a paper title).
  string content = 4;
  float score = 5;
}
